Recent topic UI

amanagr opened the pull request:
Events:
- [x] Topic Mute Updates
- [ ] Stream Mute Updates https://chat.zulip.org/#narrow/stream/6-frontend/topic/.236605.20%28Recent.20Topics%29/near/854026
- [x] Topic Edits
- [x] Stream Edits
- [ ] Stream Deletion

 * Should deleted streams be displayed in recent topics?  They are currently displayed as 
`DEACTIVATED:"Stream Name"`, we can work with the formatting to display it as
 `Stream Name (Deleted)`


Features:
- [x] Keyboard Navigation 
- [x] Mute/Unmute Topic 
- [x] Mark Topic as read
- [x] Display recent users
- [x] Open via hotkey
- [x] Show only starred topics
- [x] Add search bar for searching topics


Other UI things todo:
- [x] Only update components that are required
- [x] Make headers fixed


Followup things to do:
- [ ] We should use sortable.js to make the columns sortable by clicking the headings.
- [ ] Starred messages (problem: All message that are starred might not be initially loaded)
- [ ] Visited Topics (track data of visited topics - https://chat.zulip.org/#narrow/stream/101-design/topic/recent.20topics.20use.20cases)
- [ ] Add a keybind for cycling through recent topics. User should be able to specify which combination of filters to cycle through.
- [ ] follow/unfollow a stream+topic pair for notifications (#14555 adds backend for this)
- [ ] Add announcement section
- [ ] Mark topic as unread "Action" button

Current Discussion Link - https://chat.zulip.org/#narrow/stream/6-frontend/topic/.236605.20(Recent.20Topics)
Relevant issues: #6605, #4271 and #12309 (this issue should require a different PR however).
Started with rebase of #7714.

Thanks @aero31aero for helping me get started with this 👍

amanagr commented on static/js/recent_topics.js:
```diff
@@ -68,10 +81,88 @@ exports.process_topic_edit = function (old_stream_id, old_topic, new_topic, new_
+    return $("#" + $.escapeSelector(topic_key));
+}
+
+exports.inplace_rerender = function (topic_key) {
```
Iplace rerender function Review requested.

amanagr commented on static/templates/recent_topic_row.hbs:
```diff
@@ -0,0 +1,29 @@
+before and after a topic. This avoids
+a full rerender of the UI.
+--}}
+<tr class="empty_topic"></tr>
```
UI part of inplace logic.

amanagr replied:
I changed the to approach to using before and after jquery function as they are cleary better.

amanagr commented:
@timabbott this is mostly done, I need to correct linting errors and add tests. I would love feedback on UI and improving mechanisms/implementation.

timabbott commented:
A few thoughts on the UI, none of which are a big deal:
* s/Active Users/Participants/ in the headings.
* We should decorate the streams with #/lock/globe?
* The "Unread" column with checkmarks for read topics doesn't make sense to me; might want to look for a cleaner way to do it.
* We should use `sortable.js` to make the columns sortable by clicking the headings.

timabbott commented on static/js/recent_topics.js:
```diff
@@ -32,6 +42,9 @@ exports.process_message = function (msg) {
     topic_data.muted = topic_data.muted || muting.is_topic_muted(msg.stream_id, msg.topic);
+    if (initialized) {
+        exports.inplace_rerender(key);
+    }
```
This might end up being bad when processing 1K messages in a batch from the server.  I think what we probably want is to check the length of `messages`; if it's <5, we can do the inplace thing, otherwise, we should just do a full rerender.

amanagr replied:
Done

timabbott commented on static/js/recent_topics.js:
```diff
@@ -68,10 +81,88 @@ exports.process_topic_edit = function (old_stream_id, old_topic, new_topic, new_
+}
+
+function get_topic_row(topic_key) {
+    return $("#" + $.escapeSelector(topic_key));
```
Hmm, I'm not sure this is a good naming scheme; we should at least have an example in a comment.  Is this `escape("stream_id:topic_id")`?

amanagr replied:
Yes, I will try to make it more clear.

timabbott commented on static/templates/recent_topic_row.hbs:
```diff
@@ -0,0 +1,22 @@
+<tr id="{{stream_id}}:{{topic}}">
```
Let's at least make it `recent_topic:{{stream_id}}:{{topic}}` to namespace these IDs.

amanagr replied:
Done

timabbott commented on static/styles/recent_topics.scss:
```diff
@@ -56,5 +56,18 @@
+            // header text "Unread" so, that final width
+            // is set to header's width.
+            width: 1px;
+        }
```
Why do we need this `width` at all?

amanagr replied:
I wanted Unread column to take minimum space, since little space is required here and so that we have more space for other columns (like topic)

timabbott commented on static/js/recent_topics.js:
```diff
@@ -1,10 +1,20 @@
+        // later on.
+        exports.complete_rerender();
+        initialized = true;
+    }
```
I think perhaps we wouldn't need this `initialized` flag if we did the rendering logic for when to do complete vs. not change I suggest below.

amanagr replied:
Yes, removed.

timabbott commented on static/js/unread_ops.js:
```diff
@@ -15,6 +15,7 @@ function process_newly_read_message(message, options) {
         message_list.narrowed.show_message_as_read(message, options);
     }
     notifications.close_notification(message);
+    recent_topics.update_topic_unread_count(message);
```
Just to make sure I understand this -- this is the way we trigger a rerender of the recent topics row after the new message arrives?  (And the data structure was already updated?)

amanagr replied:
We need to update recent topics is many cases:
* New message arrives via event (also updates unread count in the same call)  - we do process_messages and rerender.
* We fetch messages in bulk manually - we do process_messages and rerender.
* A message is read -we update recent_topics by`update_topic_unread_count(message);

timabbott commented on static/js/recent_topics.js:
```diff
@@ -41,20 +40,15 @@ exports.process_message = function (msg) {
-    }
-    const topic_data = topics.get(stream_id + ":" + topic);
-    topic_data.muted = is_muted;
+    get_topic_row(key).toggle();
```
Are you sure about moving to a `.toggle()` system?  It's harder to read and not safe against getting duplicates of an event...

timabbott replied:
I prefer using e.g. `.toggle(!is_muted)` to just set it to the desired state, rather than having a state machine where we could get a parity error.

amanagr replied:
OK! Switched to this convention.

timabbott commented on static/templates/recent_topic_row.hbs:
```diff
@@ -17,6 +17,18 @@
     <td class='recent_topic_users'>
+        <ul class="recent_avatars">
+            {{#if count_senders}}
+            <li class="recent_avatars__item">
```
Why the `__` naming here?

amanagr replied:
Yeah, I forgot to refactor this section, done now. (This is from the tutorial I followed to implement this)

timabbott commented on static/styles/recent_topics.scss:
```diff
@@ -69,5 +69,47 @@
+            padding: 0px 0px 0px 0px;
+            position: relative;
+            width: 15px; /* Forces flex items to be smaller than their contents. */
+        }
```
These feel too small.  Maybe 24px?  

Do we need to repeat the sizing here in addition to in the next block?

amanagr replied:
I think @aero31aero did a better version of this:
![image](https://user-images.githubusercontent.com/25124304/82914302-f867a000-9f8c-11ea-9a93-48ec79de992e.png)
@aero31aero  Can you send the link to your commit for this, I will cherry-pick it here at the end. (or we can do it once an initial UI for this gets merged?

timabbott replied:
We should never display avatars as circles; it looks bad with some images (E.g. bot avatars which are usually product logos).  But otherwise that lgtm.

timabbott commented on static/js/recent_topics.js:
```diff
@@ -2,6 +2,7 @@ const render_recent_topics_body = require('../templates/recent_topics_table.hbs'
 const render_recent_topic_row = require('../templates/recent_topic_row.hbs');
 const topics = new Map(); // Key is stream-id:topic.
 let initialized = false;
+const MAX_AVATAR = 3;  // Number of avatars to display
```
I think we want to show 4 (or 3 + the count block); is that what this does?  Maybe worth making the comment on its own line and explain this more clearly.

amanagr replied:
Ok, Done!

timabbott commented on static/js/recent_topics.js:
```diff
@@ -168,6 +170,32 @@ exports.update_topic_unread_count = function (message) {
+        star_icon.attr('class', 'star fa on_hover_topic_star fa-star');
+    } else {
+        star_icon.attr('class', 'star fa on_hover_topic_star fa-star-o empty-star');
+    }
```
Why not write this with `addClass/removeClass` and avoid duplicating the list of 4 classes that are fixed?

amanagr replied:
I updated it to better handle fixed classes. I want to keep it this was as it is easy to read and is robust.

timabbott commented on static/js/recent_topics.js:
```diff
@@ -168,6 +170,32 @@ exports.update_topic_unread_count = function (message) {
 }
 
+exports.toggle_star_topic = function (stream_id, topic) {
+    const key = stream_id + ":" + topic;
```
Let's add a comment explaining here that we're just starring the last message on the topic even though this is a somewhat weird choice.

amanagr replied:
Done.

timabbott commented on static/js/starred_messages.js:
```diff
@@ -34,6 +34,27 @@ exports.get_starred_msg_ids = function () {
+        }
+    };
+    return false;
+}
```
I think this logic is unsound, in that messages in  `starred_messages` aren't necessarily present in `message_store`.  

Maybe let's punt the "starred messages" commit for now?

amanagr replied:
Let me try and see if we can fetch all the starred messages first. I believe folk have around 100-200 starred messages at max, shouldn't be too much.

timabbott replied:
That's a rabbit hole that isn't worth adding as a dependency of this big change; just keep it on your TODO list for future ideas.

timabbott commented on static/js/recent_topics.js:
```diff
@@ -204,6 +230,18 @@ exports.complete_rerender = function () {
+            .find('[data-filter="all"]')
+            .addClass('btn-recent-selected');
+    } else {
+        filters.forEach(function (filter) {
```
We're trying to avoid `forEach` loops.

amanagr replied:
Done. sorry for this.

timabbott commented on static/js/recent_topics.js:
```diff
@@ -87,7 +88,18 @@ function format_topic(topic_data) {
+    }
+    if (!topic_data.participated && filters.has('participated')) {
+        hidden = true;
+    }
```
You might want to extract a `should_show_row(stream_id, topic, filters)` that can just return the value used as `hidden` here; that's be more readable.

amanagr replied:
I think it would lead to repeated computation of things like `is_starred`, `unread_count` or `participated`, or would have to returned from the extracted function.

timabbott commented on static/js/click_handlers.js:
```diff
@@ -362,6 +362,25 @@ exports.initialize = function () {
+    // Search for all table rows (this combines stream & topic names)
+    $('body').on('keyup', '#recent_topics_search', _.debounce(function () {
+        const $rows = $('.recent_topics_table tr').slice(1);
+        const val = '^(?=.*\\b' + $.trim($(this).val()).split(/\s+/).join('\\b)(?=.*\\b') + ').*$';
```
I find this line really hard to read.

amanagr replied:
OK! Made this more verbose.

timabbott commented:
I posted a batch of comments.  Overall, this is really promising, and I hope we can start merging chunks in the next couple days. (We'll have only the `t` UI to access it at first).

amanagr commented:
s/Active Users/Participants/ in the headings. DONE

We should decorate the streams with #/lock/globe? DONE

> The "Unread" column with checkmarks for read topics doesn't make sense to me; might want to look for a cleaner way to do it.

 Changed it to `check-circle` for now, which is the same as mark all msgs read icon. In future, we might want to remove the unread column and just show the counts inside the topic column.

> We should use sortable.js to make the columns sortable by clicking the headings.
Added to TODO comment at top of this PR. (Coudn't get it working according to the docs, will require me digging into this. )

> Do we want to call it "RECENT TOPICS (BETA)" in the heading? Maybe it's good enough that it'll be undiscoverable at first.

I think we should show it as beta feature for next release?

(didn't push anything for now)

timabbott commented:
So, I'd like to integrate this to master so that folks can iterate on it, without making it available anywhere other than via the `t` hotkey.  Can you tweak things to arrange that (skipped "starred messages") and fix the node tests?  I think that'll put us in a good position to do a bunch of smaller PRs fixing individual minor issues while it's convenient for folks to use this on chat.zulip.org.

amanagr commented:
@timabbott I will add new commits to the end, unless they are bugfixes, only reason being the fear of breaking things during a rebase.

timabbott commented:
I'm deploying the latest version of this to chat.zulip.org for testing.

timabbott commented:
This is great work @amanagr!  I did a thorough read of this, and I'm happy to merge it; though there's a list of about a dozen follow-up fixes below (mostly on code clarity) we should do.  I just think it'll be a better use of time to do them in a new PR rather than my potentially rereading any of these commits to re-review things.

Here's a list of small things that should be fixed in a near-term follow-up PR:
* For `exports.reify_message_id_if_available`, we should probably add a comment to make clear that we don't need to do anything unless the message was the last message in the target topic.  I think we could also optimize this massively since we should know what stream/topic the message to be reified was sent to (if nothing else, via `message_store`), so we should be able to just loop up the one topic rather than doing a loop.
* `count_senders` should be called `other_senders_count` or similar, to reflect it's not the total, it's the total - 4.
* There's a `word boundry groups` typo -- should be "boundary".
* `is_topic_hidden` might benefit from being optimized to only do the calls to other modules when necessary.  (And you don't need the `else` clauses).  I think `is_topic_muted` could be expensive in a loop if the user has hundreds of muted topics, so it'd be best to avoid that if the filters mean that detail is irrelevant because the topic has no unreads.  
* In `is_topic_hidden` / the filters logic, the `muted` filter should probably be called "include_muted", given that's what it does?  Or does it only show muted topics?
* `is_topic_hidden` should probably be called `filters_should_hide_topic`; the current name makes it sound like it's checking the current UI state, not the intended UI state based on data.  Which makes this block read very strangely:
```
+    if (is_topic_hidden(topic_data)) {
+        topic_row.hide();
+    } else {
+        topic_row.show();
+    }
```
* Rather than computing `topic_key` values in many places, you should write a 1-line function `get_topic_key(stream_id, topic)`.  You'll end up wanting to make it case-insensitive, BTW, and in any case, the deduplication will help with readability.  We probably also want to name the variable `topic_lower` when used in places like the element IDs in the DOM (or maybe just pass `topic_key` into the template rendering.
* I'm not sure about the name/interface for `exports.set_filter`; it's not totally clear what it does in the first few cases.  We should add comments or rename or change the interface; not sure which is best.
* The `Prepare Header` comment should be in sentence case ("Prepare header") like all other strings in Zulip.
* Why does `show_selected_filters` need to be called just after rendering?  I guess it avoids some conditionals in the handlebars template?
* `tableFixHead` CSS class name doesn't match our style.

It'd be great to do a quick PR cleaning out that batch of issues, which I think should all be straightforward, so we don't need to think about them further.

Following those, there's a few larger issues we'll likely want to discuss longer-term:
* We need to decide where this UI should live.  I had been imagining a "Recent topics" button below "Starred messages" in the left sidebar.
* We may want to include (group) private message conversations here as well, rebranding it "Recent conversations".
* The big structural issue is that the definition of "recent" is highly variable and unclear in a structural way.  And even if we were to fetch deeper data from the server, we would need to take a lot of care to have details like the `sender_id`s available, since queries to `message_store` will fail if we start having data sources that aren't `message_store` itself.

amanagr commented:
Thanks Tim, I will follow up on these.
